docs/guide/api-environment-frameworks.md

```diff
@@ -86,30 +86,85 @@ app.use('*', async (req, res, next) => {
+
+:::info `FetchableDevEnvironment` proposal
+
+The initial proposal had a `run` method on the `DevEnvironment` class that would allow consumers to invoke an import on the runner side by using the `transport` option. During our testing we found out that the API was not universal enough to start recommending it. At the moment, we are looking for feedback on [the `FetchableDevEnvironment` proposal](https://github.com/vitejs/vite/discussions/18191).
+
+:::
+
+`RunnableDevEnvironment` has a `runner.import` function that returns the value of the module. But this function is not available in the raw `DevEnvironment` and requires the code using the Vite's APIs and the user modules to be decoupled.
+
+For example, the following example uses the value of the user module from the code using the Vite's APIs:
 
 ```ts
-const vite = createServer()
-const routes = collectRoutes()
+// code using the Vite's APIs
+import { createServer } from 'vite'
+
+const server = createServer()
+const ssrEnvironment = server.environment.ssr
+const input = {}
+
+const { createHandler } = await ssrEnvironment.runner.import('./entrypoint.js')
+const handler = createHandler(input)
+const response = handler(new Request('/'))
 
-const { processRoutes } = await vite.ssrLoadModule('internal:routes-processor')
-processRoutes(routes)
+// -------------------------------------
+// ./entrypoint.js
+export function createHandler(input) {
+  return function handler(req) {
+    return new Response('hello')
+  }
+}
 ```
 
-This makes it impossible to run user code in the same way it might run in production (for example, on the edge) because the server state and user state are coupled. So instead, we recommend using virtual modules to import the state and process it inside the user module:
+If your code can run in the same runtime with the user modules (= does not rely on Node.js specific APIs), you can use a virtual module so that you don't need to access the value from the code using the Vite's APIs.
 
 ```ts
-// this code runs on another machine or in another thread
+// code using the Vite's APIs
+import { createServer } from 'vite'
 
-import { runner } from './ssr-module-runner.js'
-import { processRoutes } from './routes-processor.js'
+const server = createServer({
+  plugins: [
+    // a plugin that handles `virtual:entrypoint`
+    {
+      name: 'virtual-module',
+      /* plugin implementation */
+    },
+  ],
+})
+const ssrEnvironment = server.environment.ssr
+const input = {}
+
+// use exposed functions by each environment factories that runs the code
+// check for each environment factories what they provide
+if (ssrEnvironment instanceof RunnableDevEnvironment) {
+  ssrEnvironment.runner.import('virtual:entrypoint')
+} else if (ssrEnvironment instanceof CloudflareDevEnvironment) {
+  ssrEnvironment.evaluate('virtual:entrypoint')
```

Comment:
maybe it's better to have a fetchable API here? The `evaluate` and `cloudflare` are a bit confusing here - especially because `cloudflare` is a fetchable environment

Reply:
I've pushed a commit that replaces them with `CustomDevEnvironment` and `.runEntrypoint` so that it doesn't look like it can satisfy the `FetchableDevEnvironment` interface.